Set-output-expression: Enable setting of a prior to an output node

HenrikMettler opened the pull request:
In some use cases it is nice to set (one of) the output(s) to a prior. One way of doing this is by inserting a dna sequence corresponding to the prior at the first internal node(s). `slice_dna` gives this possibility, without requiring the user to set the full dna. 

Marked as draft, since test coverage is minimal and code not optimized. 

(Note: I also looked at implementing the prior from a function/sympy string, however this seems quite tedious, especially if the prior contains more than 1 operation. This implementation at least circumvents the need to set the full dna :) )

closes #317 (by circumventing the issue)

jakobj commented on cgp/genome.py:
```diff
@@ -270,6 +269,57 @@ def randomize(self, rng: np.random.RandomState) -> None:
         # accept generated dna if it is valid
         self.dna = dna
 
+    def slice_dna(self, dna_insert: List[int], output_node_idx: int = 0):
```
maybe `slice_dna` -> `splice`?

jakobj replied:
maybe also also allow providing a starting point which by default is 0?

jakobj commented on cgp/genome.py:
```diff
@@ -394,10 +444,12 @@ def _replace_invalid_address_alleles(self, dna: List[int], rng: np.random.Random
 
     def _get_addable_nodes(
-        self, node_dependencies: Dict[int, Set[int]], used_node_indices: List[int] = [],
+        self,
```
are these changes necessary for this PR?

HenrikMettler replied:
No. I think there is an error in my pre-commit hooks, will change this and revert these changes

jakobj commented on cgp/genome.py:
```diff
@@ -413,7 +465,7 @@ def _get_region(self, region_idx: int, dna: List[int]) -> List[int]:
 
     def _determine_node_dependencies(self) -> Dict[int, Set[int]]:
-        """ Determines the set of node indices on which each node depends.
+        """Determines the set of node indices on which each node depends.
```
are these changes necessary for this PR?

jakobj commented on test/test_genome.py:
```diff
@@ -13,7 +14,11 @@ def test_check_dna_consistency():
     primitives = (cgp.Add,)
     genome = cgp.Genome(
-        params["n_inputs"], params["n_outputs"], params["n_columns"], params["n_rows"], primitives,
+        params["n_inputs"],
```
are these changes necessary for this PR?

jakobj commented on test/test_genome.py:
```diff
@@ -210,7 +215,11 @@ def test_region_iterators():
     primitives = (cgp.Add,)
     genome = cgp.Genome(
-        params["n_inputs"], params["n_outputs"], params["n_columns"], params["n_rows"], primitives,
+        params["n_inputs"],
```
are these changes necessary for this PR?

jakobj commented on test/test_genome.py:
```diff
@@ -766,3 +775,16 @@ def test_ncolumns_zero(rng):
+
+def test_slice_dna(genome_params, rng):
+
+    genome = cgp.Genome(**genome_params)
```
just to make sure to not get false positives you could do this 10 times

HenrikMettler replied:
I wasn't fully sure what to do (the logic changed since you wrote that message) - but I have added a few more (complicated) expressions for the output (in `test_set_expression_for_output`)

jakobj reviewed (changes requested):
awesome, thanks a lot for working on this! :rocket: 

i've left a few comments, in particular i wasn't sure whether all changes are relevant for this PR. please remove those which are not.

jakobj commented:
ping @HenrikMettler

HenrikMettler commented on .pre-commit-config.yaml:
```diff
@@ -1,6 +1,6 @@
 repos:
 -   repo: https://github.com/ambv/black
-    rev: stable
+    rev: 19.10b0
```
Has to be the same as in extra_requirements, otherwise there seems to be a formatting conflict.

HenrikMettler replied:
(move to seperate PR)

HenrikMettler commented on cgp/genome.py:
```diff
@@ -270,6 +269,68 @@ def randomize(self, rng: np.random.RandomState) -> None:
+        None
+        """
+
+        # check that dna is initialized
+        if len(self.dna) == 0:
+            raise AssertionError(
+                "Can only splice genome with initialized dna, " "use genome.randomize first"
+            )
```
Can be removed, since checked in self.dna

HenrikMettler commented on cgp/genome.py:
```diff
@@ -270,6 +269,68 @@ def randomize(self, rng: np.random.RandomState) -> None:
+                "Choose smaller internal_start_node or idx"
+            )
+
+        dna = self._dna.copy()
```
```suggestion
        dna = self.dna
```

HenrikMettler commented on cgp/genome.py:
```diff
@@ -270,6 +269,68 @@ def randomize(self, rng: np.random.RandomState) -> None:
+            )
+        # check that dna insert is a multiple of region length
+        if len(dna_insert) % (self._length_per_region) != 0:
+            raise AssertionError("Dna insert is not of proper length")
```
```suggestion
            raise ValueError("Dna insert is not of proper length")
```

HenrikMettler commented on cgp/genome.py:
```diff
@@ -270,6 +269,68 @@ def randomize(self, rng: np.random.RandomState) -> None:
+        # check that output_node_idx is valid
+        if output_node_idx > self._n_outputs - 1:
+            raise ValueError("Output node idx too high, choose wrt to n_outputs")
+
```
add test that `output_node_idx` is larger than zero.

HenrikMettler commented:
I changed the logic a bit: The two parts of the previous `splice_dna` function are now split up into two separate functions (for splicing and changing the output address ). What was previously splice_dna is now called `set_prior` to better reflect what the function actually does. This function calls the other two functions once each.

jakobj commented on cgp/cartesian_graph.py:
```diff
@@ -63,13 +63,11 @@ def __repr__(self) -> str:
     def print_active_nodes(self) -> str:
-        """Print a representation of all active nodes in the graph.
-        """
+        """Print a representation of all active nodes in the graph."""
```
please revert all formatting changes in this PR

jakobj commented on cgp/genome.py:
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
         # accept generated dna if it is valid
         self.dna = dna
 
+    def splice_dna(self, dna_insert: List[int], internal_start_node: int = 0) -> int:
```
```suggestion
    def splice_dna(self, new_dna: List[int], internal_start_node: int = 0) -> int:
```

jakobj replied:
also in the following

jakobj commented on cgp/genome.py:
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
+    def splice_dna(self, dna_insert: List[int], internal_start_node: int = 0) -> int:
+        """Splice the dna.
+
+        Replaces a segment of the dna starting from internal start node with the dna insert
```
```suggestion
        Replaces a segment of the dna starting from internal start node with new dna.
```

jakobj commented on cgp/genome.py:
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
+        Parameters
+        ----------
+        dna_insert: List[int]
+            dna segment to be inserted.
```
```suggestion
            New dna to be inserted.
```

jakobj commented on cgp/genome.py:
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
+        ----------
+        dna_insert: List[int]
+            dna segment to be inserted.
+        internal_start_node: int
```
```suggestion
        internal_start_node: int, optional
```

jakobj commented on cgp/genome.py:
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
+        dna_insert: List[int]
+            dna segment to be inserted.
+        internal_start_node: int
+            idx of the internal node, where the insert starts.
```
```suggestion
            Index of the internal node, where the insert starts.
```

jakobj commented on cgp/genome.py:
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
+
+        Returns
+        ----------
+        address_last_inserted_node: int
+            address of the last node inserted.
```
```suggestion
        int
            Address of the last node for which the dna was changed.
```

jakobj commented on cgp/genome.py:
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
+            address of the last node inserted.
+        """
+        if len(dna_insert) % self._length_per_region != 0 or len(dna_insert) == 0:
+            raise ValueError("Length of inserted dna must be a multiple of region length")
```
```suggestion
            raise ValueError(f"Length of inserted dna must be a multiple of region length ({self._length_per_region})")
```

provides a bit more info to the user

jakobj commented on cgp/genome.py:
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
+        n_inserted_nodes = len(dna_insert) // self._length_per_region
+
+        if internal_start_node + n_inserted_nodes >= self._n_hidden:
+            raise ValueError("Inserted dna too long")
```
```suggestion
            raise ValueError("New dna too long")
```

jakobj commented on cgp/genome.py:
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
+
+        self.dna = dna
+
+        address_last_inserted_node = self._n_inputs + internal_start_node + n_inserted_nodes - 1
```
the "address" is just the same as the index, right? so we could drop the "address_" prefix here, it doesn't provide more information. we also say "internal_start_node" and not "internal_start_node_address"

jakobj commented on cgp/genome.py:
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
+
+        return address_last_inserted_node
+
+    def change_output_address(self, address_node: int, output_node_idx: int = 0):
```
```suggestion
    def change_address_gene_of_output_node(self, new_address: int, output_node_idx: int = 0):
```

jakobj commented on cgp/genome.py:
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
+        return address_last_inserted_node
+
+    def change_output_address(self, address_node: int, output_node_idx: int = 0):
+        """Change the address of an output gene
```
```suggestion
        """Change the address gene of an output node.
```

jakobj commented on cgp/genome.py:
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
+        address_node: int
+            Address of the node from which the output should read
+        output_node_idx: int
+            If defined, idx of the output node for which the address node is changed.
```
```suggestion
            Index of the output node for which the address node is changed.
```
are these indices relative (0 is the first output) or absolute (0 is the first input)? mention explicitly here please

jakobj commented on cgp/genome.py:
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
+        None
+        """
+        if address_node < 0 or address_node > self._n_hidden:
+            raise ValueError("Address node non valid")
```
```suggestion
            raise ValueError("New address not valid; must be the index of a hidden node")
```

jakobj commented on cgp/genome.py:
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
+        if address_node < 0 or address_node > self._n_hidden:
+            raise ValueError("Address node non valid")
+        if output_node_idx < 0:
+            raise ValueError("Output node idx must be non-negative")
```
```suggestion
            raise ValueError("Output node index must be non-negative")
```

jakobj commented on cgp/genome.py:
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
+        if output_node_idx < 0:
+            raise ValueError("Output node idx must be non-negative")
+        if output_node_idx > self._n_outputs - 1:
+            raise ValueError("Output node idx must be smaller than number of outputs")
```
```suggestion
            raise ValueError("Output node index must be smaller than number of outputs")
```

jakobj commented on cgp/genome.py:
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
+        dna[idx_output_address_gene] = address_node
+        self.dna = dna
+
+    def set_prior(
```
if not sure whether "prior" is the best name. can't we just say "set_expression_for_output" or something like that?

jakobj commented on cgp/genome.py:
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
+    ):
+        """Set a prior for one output node
+
+        Replaces internal (hidden) nodes with user defined insert, starting from desired internal
+        node. Sets a definable output node to read from the last node of the insert.
```
```suggestion
            Replaces part of the dna with user defined values starting from the specified internal
            node. Sets the specified output node to read from the last node for which dna was changed.
```

jakobj commented on cgp/genome.py:
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
+        dna_insert: List[int]
+            dna segment to be inserted at the first internal (hidden) nodes.
+        internal_start_node: int
+            idx of the internal node, where the insert starts.
```
```suggestion
            index of the internal node, where the insert starts.
```

jakobj commented on cgp/genome.py:
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
+            idx of the internal node, where the insert starts.
+            Defaults to 0.
+        output_node_idx: int
+            If defined, idx of the output node which will read the last node of the insert.
```
```suggestion
            Index of the output node which will read the last node of the insert.
```

jakobj commented on cgp/genome.py:
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
+        None
+        """
+
+        # insert new dna at desired location
```
```suggestion
```
your code and function names are so clear, no need for these comments

jakobj commented on cgp/genome.py:
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
+        # insert new dna at desired location
+        address_last_inserted_node = self.splice_dna(dna_insert, internal_start_node)
+
+        # make selected output node read from last inserted node
```
```suggestion
```

jakobj commented on test/test_genome.py:
```diff
@@ -766,3 +767,77 @@ def test_ncolumns_zero(rng):
+    genome = cgp.Genome(**genome_params)
+
+    with pytest.raises(ValueError):
+        [] = genome.splice_dna(dna_insert=[1, 2, 3])
```
```suggestion
        genome.splice_dna(dna_insert=[1, 2, 3])
```
not sure what this does, but if you don't want to use the return value of a function you just don't assign it a name

jakobj reviewed (changes requested):
yes, this split is nice, makes the code super easy to follow, thanks! :ok_hand: 

added some comments, mostly on comments and variable names

jakobj commented:
ah, and also please rebase on the current master, there are some conflicts with the main branch.

jakobj commented:
ping @HenrikMettler

HenrikMettler commented:
ping @jakobj

jakobj reviewed (changes requested):
almost perfect ;)

i've added a few more comments that should be addressed before merging.

the tests look great, i also can't interpret my old comment any more.

HenrikMettler commented on cgp/genome.py:
```diff
@@ -270,6 +270,110 @@ def randomize(self, rng: np.random.RandomState) -> None:
+
+        Returns
+        ----------
+        address_last_inserted_node: int
```
I guess you meant `last_inserted_node: int` ?

jakobj reviewed (approved):
awesome! :rocket: :+1: and merging.

jakobj commented:
ah, no merging yet: could you please squash the commits? :)

jakobj commented:
awesome!
